sklearn/preprocessing/_encoders.py

```diff
@@ -41,18 +41,38 @@ def _check_X(self, X):
           not do that)
 
         """
+        if hasattr(X, 'iloc') and getattr(X, 'ndim', 0) == 2:
+            # only if pandas dataframe (not an array, sequence etc...)
+            # do not check full X,
+            # it should be checked column by column (using _check_X_feature())
+            return X
+
         X_temp = check_array(X, dtype=None)
         if not hasattr(X, 'dtype') and np.issubdtype(X_temp.dtype, np.str_):
             X = check_array(X, dtype=np.object)
         else:
             X = X_temp
 
-        if X.dtype == np.dtype('object'):
+        return X
```

Comment:
@jorisvandenbossche had the thought that we could just return a list of columns in all cases from this. That way we could work consistently with the different types.

Reply:
_check_X() now returns the list of features... 
should we also keep the dtypes of features if X is a list? But then we would have to check specifically if X is a list and work with it (instead of directly converting it to array-like object) maybe too messy?